projectcontenthandler.cpp

```diff
@@ -0,0 +1,179 @@
+{
+    m_tree->treeAboutToChange();
+
+    m_storage->addProject(projectName);
+    auto node = ModelTreeNode::ModelTreeNodePtr(new ModelTreeNode(projectName));
+    m_tree->addProjectNode(node);
+
+    allProcessingFinished();
+}
+
+void ProjectContentHandler::handleProjectRemoved(const QString &projectName)
+{
+    m_tree->treeAboutToChange();
+
+    m_tree->removeProjectNode(projectName);
+    m_storage->removeProject(projectName);
+
+    allProcessingFinished();
+}
+
+void ProjectContentHandler::handleFileListChanged(const QString &projectName, const QStringList &fileList)
+{
+    int storedFilesCnt = m_tree->getProjectFilesCnt(projectName);
+
+    int currentFilesCnt = fileList.count();
+
+    m_tree->treeAboutToChange();
+    if (currentFilesCnt > storedFilesCnt)
+        handleFilesAdded(projectName, fileList);
+    else if (currentFilesCnt < storedFilesCnt)
+        handleFilesRemoved(projectName, fileList);
+}
+
+void ProjectContentHandler::handleFileContentChanged(const QString &path, const QString &content, int revision)
+{
+    m_tree->treeAboutToChange();
+
+    QStringList projects = m_storage->getProjectsForFile(path);
+    for (const QString projectName : projects) {
+        DocumentProcessor *docProcessor = new DocumentProcessor(projectName);
+
+        QList<std::shared_ptr<ParsedDocument> > files = m_storage->getFilesFromProject(projectName);
+        foreach (const std::shared_ptr<ParsedDocument> &file, files) {
+            if (file->source().getPath() == path) {
+                docProcessor->addToRun(content, path, revision);
+            } else {
+                docProcessor->addToRun(file->source().getContent(),
+                                       file->source().getPath(),
+                                       file->source().getRevision());
+            }
+        }
+
+        connect(docProcessor, &DocumentProcessor::processingFinished,
+                [this, docProcessor](){docProcessor->deleteLater();});
+
+        connect(docProcessor, &DocumentProcessor::processingFinished,
+                this, &ProjectContentHandler::onFilesProcessingFinished);
+
+        docProcessor->run();
+
+        m_projectsChanged++;
+    }
+}
+
+void ProjectContentHandler::handleFilesAdded(const QString &projectName, const QStringList &filePaths)
+{
+    foreach (const QString &path, filePaths) {
+        if (m_tree->getNodeForFilepathFromProject(projectName, path) != nullptr)
+            continue;
+
+        auto node = ModelTreeNode::ModelTreeNodePtr(new ModelTreeNode(path));
+        m_tree->addNodeToProject(projectName, node);
+    }
+
+    processFiles(projectName, filePaths);
+}
+
+void ProjectContentHandler::handleFilesRemoved(const QString &projectName, const QStringList &filePaths)
+{
+    m_tree->removeStaleNodesFromProject(projectName, filePaths);
+    m_storage->cleanProject(projectName);
+
+    // FIXME: if all the files were removed from project, than asn1scc daemon will crash
```

Comment:
- "// FIXME: if all the files were removed from project, than asn1scc daemon will crash" - did not create JIRA path for it, but resolved the issue instead.